File: Purple_3.cs
using System;
using System.Linq;

namespace Lab_6
{
    public class Purple_3
    {
        public struct Participant
        {
            private string _firstName;
            private string _lastName;
            private double[] _marks;
            private int[] _places;
            private int _judgeCount;

            public string Name => _firstName;
            public string Surname => _lastName;
            public double[] Marks
            {
                get
                {
                    if (_marks == null)
                        return null;
                    double[] copy = new double[_marks.Length];
                    Array.Copy(_marks, copy, _marks.Length);
                    return copy;
                }
            }
            public int[] Places
            {
                get
                {
                    if (_places == null)
                        return null;
                    int[] copy = new int[_places.Length];
                    Array.Copy(_places, copy, _places.Length);
                    return copy;
                }
            }
            public int Score => _places == null ? 0 : _places.Sum();

            private int TopPlace => _places == null ? 0 : _places.Min();
            private double TotalSum => _marks == null ? 0 : _marks.Sum();

            public Participant(string name, string surname)
            {
                _firstName = name;
                _lastName = surname;
                _marks = new double[7];
                _places = new int[7];
                _judgeCount = 0;
            }

            public void Evaluate(double result)
            {
                if (_judgeCount < 7)
                {
                    _marks[_judgeCount] = result;
                    _judgeCount++;
                }
            }

            public static void SetPlaces(Participant[] participants)
            {
                if (participants == null) return;
                
                for (int i = 0; i < 7; i++)
                {
                    Participant[] arr = participants
                        .OrderByDescending(p => p._marks[i])
                        .ThenBy(p => (p._places != null && p._places.Length > 0) ? p._places[p._places.Length - 1] : int.MaxValue)
                        .ToArray();
                    
                    Array.Copy(arr, participants, participants.Length);
                    
                    for (int j = 0; j < participants.Length; j++)
                    {
                        participants[j]._places[i] = j + 1;
                    }
                }
                Participant[] finalSorted = participants
                    .OrderBy(p => (p._places != null && p._places.Length > 0) ? p._places[p._places.Length - 1] : int.MaxValue)
                    .ToArray();
                Array.Copy(finalSorted, participants, participants.Length);
            }

            public static void Sort(Participant[] array)
            {
                if (array == null) return;
                var sorted = array
                    .OrderBy(p => p.Score)
                    .ThenBy(p => p.TopPlace)
                    .ThenByDescending(p => p.TotalSum)
                    .ToArray();
                Array.Copy(sorted, array, sorted.Length);
            }

            public void Print()
            {
                Console.WriteLine($"{_firstName} {_lastName}");
                Console.Write("Marks: ");
                foreach (double mark in Marks)
                {
                    Console.Write(mark + "  ");
                }
                Console.WriteLine();
                Console.Write("Places: ");
                foreach (int place in Places)
                {
                    Console.Write(place + "  ");
                }
                Console.WriteLine();
                Console.WriteLine("Total Score: " + Score);
                Console.WriteLine();
            }
        }
    }
}

File: Additional tools/Purple_3.py
input_text = """Name
Surname
Marks
Виктор
Полевой
5,93
5,44
1,20
0,28
1,57
1,86
5,89
Алиса
Козлова
1,68
3,79
3,62
2,76
4,47
4,26
5,79
Ярослав
Зайцев
2,93
3,10
5,46
4,88
3,99
4,79
5,56
Савелий
Кристиан
4,20
4,69
3,90
1,67
1,13
5,66
5,40
Алиса
Козлова
3,27
2,43
0,90
5,61
3,12
3,76
3,73
Алиса
Луговая
0,75
1,13
5,43
2,07
2,68
0,83
3,68
Александр
Петров
3,78
3,42
3,84
2,19
1,20
2,51
3,51
Мария
Смирнова
1,35
3,40
1,85
2,02
2,78
3,23
3,03
Полина
Сидорова
0,55
5,93
0,75
5,15
4,35
1,51
2,77
Татьяна
Сидорова
3,86
0,19
0,46
5,14
5,37
0,94
0,84"""

lines = input_text.split("\n")[3:]
participants = []
i = 0

while i < len(lines):
    first_name = lines[i]
    last_name = lines[i + 1]
    marks = []
    i += 2
    
    while i < len(lines) and "," in lines[i]:
        marks.append(float(lines[i].replace(",", ".")))
        i += 1
    
    participants.append((first_name, last_name, marks))

output = []
for idx, (first_name, last_name, marks) in enumerate(participants):
    output.append(f'participants3[{idx}] = new Purple_3.Participant("{first_name}", "{last_name}");')
    for mark in marks:
        output.append(f'participants3[{idx}].Evaluate({mark});')

print("\n".join(output))
